File: leetcode/Go/code/104.二叉树的最大深度/104.二叉树的最大深度.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 一路沿着最左侧或者最右侧递归查找下去, 深度检索
// 有值加1, 无值返回0
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return 1 + max(maxDepth(root.Left), maxDepth(root.Right))
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// 宽度检索
func maxDepthWithBFS(root *TreeNode) int {
	if root == nil {
		return 0
	}

	result := 0
	nodeList := make([]*TreeNode, 0)
	nodeList = append(nodeList, root)

	for len(nodeList) > 0 {
		result++

		n := len(nodeList)
		for a := 0; a < n; a++ {
			if nodeList[0].Left != nil {
				nodeList = append(nodeList, nodeList[0].Left)
			}
			if nodeList[0].Right != nil {
				nodeList = append(nodeList, nodeList[0].Right)
			}
			nodeList = nodeList[1:]
		}
	}

	return result
}

File: leetcode/Python/code/104.二叉树的最大深度/104.二叉树的最大深度.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth_with_dfs(self, root: Optional[TreeNode]) -> int:
        """深度检索"""
        # 如果没有值, 那就返回0
        if not root:
            return 0

        # 沿着左树 或者 右树, 一直找, 谁最大返回谁
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))

    def maxDepth_with_bfs(self, root: Optional[TreeNode]) -> int:
        """宽度检索"""
        if not root:
            return 0

        result = 0
        node_list = [root]
        while node_list:
            result += 1
            n = len(node_list)
            for _ in range(n):
                if node_list[0].left:
                    node_list.append(node_list[0].left)
                if node_list[0].right:
                    node_list.append(node_list[0].right)
                node_list.pop(0)

        return result
